### PR TITLE
Simplify `Use_` structs

### PR DESCRIPTION
**Pull request by haberdashPI**
After working on #22 the proliferation of `Use_` structs was really starting to bother me, and this weekend I thought of a way to simplify stuff a lot. 

The intention would be to merge this before tagging the 1.0 release, to avoid useless churn / breakage between this and #22.

This also takes some structures from `setup_tests.jl` (`TablesEq` and `ViewsEq`) and adds them to the pacakge properly. They are already used in the tests to help exercise various corner-cases I want to avoid bugs in, and they are actually probably useful outside of testing. There were a few gotcha's in getting these to work, so it seemed best to make them available.

### CODE REVIEW

File: src/StableHashTraits.jl [hunk 1]
```diff
@@ -223,113 +224,202 @@ You should return one of the following values.
...
+
+### Customizing hashes within an object
+
+Contexts can be changed not only when you call `stable_hash` but also when you hash the
+contents of a particular object. This lets you change how hashing occurs within the object.
+See the docstring of [`UseAndReplaceContext`](@ref) for details. 
+"""
+hash_method(x, context) = hash_method(x, parent_context(context))
+hash_method(x, ::Nothing) = hash_method(x)
+hash_method(::Any) = nothing
```
Comment (ssfrr) [inline src/StableHashTraits.jl:322]: remove 1-argument `hash_method` methods

**Review by ssfrr — CHANGES_REQUESTED** [review]

File: src/StableHashTraits.jl (hunk 1 above)
Comment (haberdashPI) [reply src/StableHashTraits.jl]: Thinking on this more I realized the reall problem is with method ambiguities. I've update the comments to describe this more precisely for anyone wondering why this is the way it works.

**Review by ssfrr — APPROVED** [review]
Looks great, thanks for pushing on this! A few comments and suggestions.

File: README.md [hunk 2]
```diff
@@ -21,86 +21,72 @@ struct MyType
...
-- `AbstractSet`: `Use(qualified_name, UseTransform(sort! ∘ collect))`
-- `AbstractDict`: `Use(qualified_name, UseStruct(keys => getindex, :ByName))`
-
-There are two built-in contexts that can be used to modify these default fallbacks:
-`TablesEq` and `ViewsEq`. `TablesEq` makes any table with equivalent content have the same
-hash, and `ViewsEq` makes any array or string with the same sequence of values and the same
-size have an equal hash. You can pass one or more of these as the second argument to `stable_hash`, e.g. `stable_hash(x, ViewsEq())` or `stable_hash(x, ViewsEq(TablesEq()))`.
+`write` is the same for an object that uses `WriteHash`, its hash will be the same; if the
+fields are the same for `StructHash`, the hash will be the same; etc...
+<!-- END_HASH_TRAITS -->
```
[the remark this answers is not in the document]
Comment (haberdashPI) [reply README.md:91]: I think it requires the reader to understand contexts, which are not fully explained until later in the README.

File: src/StableHashTraits.jl [hunk 3]
```diff
@@ -61,65 +152,58 @@ See also: [`StableHashTraits.hash_method`](@ref).
...
-    interior_hash = digest!(result)
-    # digest will return nothing if no objects have been added to the hash when using
-    # GenericFunHash; in this case, don't update the hash at all
-    if !isnothing(interior_hash)
-        update!(hash, copy(reinterpret(UInt8, vcat(interior_hash))))
-    end
-    return hash
+function recursive_hash!(hash_state, nested_hash_state)
+    nested_hash = digest!(nested_hash_state)
+    update!(hash_state, copy(reinterpret(UInt8, vcat(nested_hash))))
```
[the remark this answers is not in the document]
Comment (haberdashPI) [reply src/StableHashTraits.jl:164]: This was required for CRC32c. In reality I don't think we ever really want to use it, and it can still work if you pass `alg=(x, s=0x000000) -> CRC32c.crc32c(copy(x), s)`, so I will remove the copy.

File: src/StableHashTraits.jl [hunk 4]
```diff
@@ -136,216 +220,164 @@ function validate_name(str)
...
+hash_method(::Symbol, ::HashVersion{1}) = (ConstantHash(":"), WriteHash())
 function hash_method(::AbstractDict, ::HashVersion{1})
-    return Use(qualified_name, UseStruct(keys => getindex, :ByName))
+    return (FnHash(qualified_name), StructHash(keys => getindex, :ByName))
+end
+hash_method(::Tuple, ::HashVersion{1}) = (FnHash(qualified_name), IterateHash())
+hash_method(::Pair, ::HashVersion{1}) = (FnHash(qualified_name), IterateHash())
+function hash_method(::Type, ::HashVersion{1})
+    return (ConstantHash("Base.DataType"), FnHash(qualified_name))
+end
```
[the remark this answers is not in the document]
Comment (haberdashPI) [reply src/StableHashTraits.jl:374]: This is just a mistake: should be `qualified_type`

File: src/StableHashTraits.jl [hunk 5]
```diff
@@ -136,216 +220,164 @@ function validate_name(str)
...
 
-    then_h = stable_hash_helper(x, similar_hasher(hash), context, method.then)
-    return recursive_hash!(hash, then_h)
+    return hash_state
 end
 
 """
 
-    UseAndReplaceContext(method, old_context -> new_context)
+    HashAndContext(method, old_context -> new_context)
```
[the remark this answers is not in the document]
Comment (haberdashPI) [reply src/StableHashTraits.jl:263]: I think the problem with that order is that it is opposite the order the two things happen in.

i.e. first we apply `method` to the input object, and then, within the application of `method` we use `new_context`.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssfrr) [reply README.md:91]: I think it would still be good to include it. This list is a convenient place to look for "what traits are available", so I think it would be good to include `HashAndContext` even if it's a little out-of-order in the README.

File: src/StableHashTraits.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (ssfrr) [reply src/StableHashTraits.jl:268]: ah, I see. that makes sense. thanks.
